play.pokemonshowdown.com/src/panel-chat.tsx

```diff
@@ -669,6 +701,107 @@ export class ChatTextEntry extends preact.Component<{
+			} else {
+				this.tabComplete.index++;
+			}
+			if (this.tabComplete.index >= this.tabComplete.candidates?.length) this.tabComplete.index = 0;
+			if (this.tabComplete.index < 0) this.tabComplete.index = this.tabComplete.candidates?.length - 1;
+		} else {
+			// This is a new tab completion.
+			// There needs to be non-whitespace to the left of the cursor.
+			// no command prefixes either, we're testing for usernames here.
+			prefix = prefix.trim();
+
+			let m1 = /^([\s\S!/]*?)([A-Za-z0-9][^, \n]*)$/.exec(prefix);
+			let m2 = /^([\s\S!/]*?)([A-Za-z0-9][^, \n]* [^, ]*)$/.exec(prefix);
+			if (!m1 && !m2) return true;
+
+			this.tabComplete.prefix = prefix;
+			let idprefix = (m1 ? toID(m1[2]) : '');
+			let spaceprefix = (m2 ? m2[2].replace(/[^A-Za-z0-9 ]+/g, '').toLowerCase() : '');
+			let candidates: string[][] = []; // array of [candidate userid, prefix length]
+			// don't include command names in autocomplete
+			if (m2 && (m2[0] === '/' || m2[0] === '!')) spaceprefix = '';
+			for (let i in users) {
+				if (spaceprefix && users[i].replace(/[^A-Za-z0-9 ]+/g, '')
+					.toLowerCase()
+					.substr(0, spaceprefix.length) === spaceprefix) {
+					if (m2) candidates.push([i, String(m2[1].length)] as never);
+				} else if (idprefix && i.substr(0, idprefix.length) === idprefix) {
+					if (m1) candidates.push([i, m1[1].length] as never);
+				}
+			}
+			// Sort by most recent to speak in the chat, or, in the case of a tie,
+			// in alphabetical order.
+			candidates.sort((a, b) => {
+				if (a[1] !== b[1]) {
+					// shorter prefix length comes first
+					return Number([1]) - Number(b[1]);
+				}
+				let userActivity = (this.props.room as ChatRoom).userActivity;
+				let aidx = userActivity?.indexOf(a[0]) || -1;
+				let bidx = userActivity?.indexOf(b[0]) || -1;
+				if (aidx !== -1) {
+					if (bidx !== -1) {
+						return bidx - aidx;
+					}
+					return -1; // a comes first
+				} else if (bidx !== -1) {
+					return 1; // b comes first
+				}
+				return (a[0] < b[0]) ? -1 : 1; // alphabetical order
+			});
+			this.tabComplete.candidates = candidates;
+			this.tabComplete.index = 0;
+			if (!candidates.length) {
+				this.tabComplete.cursor = null;
+				return true;
+			}
+		}
+		// Substitute in the tab-completed name.
+		let candidate = this.tabComplete.candidates ? this.tabComplete.candidates[this.tabComplete.index] : '';
+		let substituteUserId = candidate[0];
+		let substituteUser = users[substituteUserId] || candidate[1];
+		if (!substituteUser) return true;
+		let name = substituteUser;
+		// Remove rank and busy charaxcters
+		name = Dex.getShortName(name).replace(/^[^a-zA-Z0-9]+|[^a-zA-Z0-9]+$/g, "");
+		let prefixIndex = candidate[1].toString().startsWith('/') ? prefix.lastIndexOf('\n') + 1 : candidate[1];
+		let fullPrefix = this.tabComplete.prefix?.substr(0, prefixIndex as number) + name;
+		this.miniedit?._setContent(fullPrefix);
+		let pos = fullPrefix.length;
+		this.setSelection(pos, pos);
```

Comment:
How did you even find the `_setContent` function??? Anyway, if a function starts with `_`, that's a good sign you're using the wrong one.

```suggestion
		this.setValue(fullPrefix, { start: fullPrefix.length, end: fullPrefix.length });
```

Reply:
lmaoo my editor autofilled it and it didnt warn me so i thought eh, it worked tho but yeah i shouldnt use it